mava/systems/sable/sebulba/rec_sable.py

```diff
@@ -0,0 +1,787 @@
+        )
+
+        return learner_state, loss_info
+
+    return learner_fn
+
+
+def learner_thread(
+    learn_fn: Callable[
+        [LearnerState, Transition, HiddenStates, HiddenStates], Tuple[LearnerState, Metrics]
+    ],
+    learner_state: LearnerState,
+    config: DictConfig,
+    eval_queue: Queue,
+    pipeline: Pipeline,
+    params_sources: Sequence[ParamsSource],
+) -> None:
+    for _ in range(config.arch.num_evaluation):
+        # Create the lists to store metrics and timings for this learning iteration.
+        metrics: List[Tuple[Dict, Dict]] = []
+        rollout_times_array: List[Dict] = []
+        learn_times: Dict[str, List[float]] = defaultdict(list)
+
+        with RecordTimeTo(learn_times["learner_time_per_eval"]):
+            for _ in range(config.system.num_updates_per_eval):
+                # Get the trajectory batch from the pipeline
+                # This is blocking so it will wait until the pipeline has data.
+                with RecordTimeTo(learn_times["rollout_get_time"]):
+                    (
+                        traj_batch,
+                        rollout_time,
+                        ep_metrics,
+                        (timestep, (prev_hstates, updated_hstates)),
+                    ) = pipeline.get(block=True)  # type: ignore
+
+                # Replace the timestep in the learner state with the latest timestep
+                # This means the learner has access to the entire trajectory as well as
+                # an additional timestep which it can use to bootstrap.
+                learner_state = learner_state._replace(timestep=timestep)
+                # Update the networks
+                with RecordTimeTo(learn_times["learning_time"]):
+                    learner_state, train_metrics = learn_fn(
+                        learner_state, traj_batch, prev_hstates, updated_hstates
+                    )
+
+                metrics.append((ep_metrics, train_metrics))
+                rollout_times_array.append(rollout_time)
+
+                # Update all the params sources so all actors can get the latest params
+                params = jax.block_until_ready(learner_state.params)
+                for source in params_sources:
+                    source.update(params)
+
+        # Pass all the metrics and  params to the main thread (evaluator) for logging and evaluation
+        ep_metrics, train_metrics = tree.map(lambda *x: np.asarray(x), *metrics)
+        rollout_times: Dict[str, NDArray] = tree.map(lambda *x: np.mean(x), *rollout_times_array)
+        timing_dict = rollout_times | learn_times
+        timing_dict = tree.map(np.mean, timing_dict, is_leaf=lambda x: isinstance(x, list))
+
+        eval_queue.put((ep_metrics, train_metrics, learner_state, timing_dict))
+
+
+def learner_setup(
+    key: chex.PRNGKey, config: DictConfig, learner_devices: List
+) -> Tuple[
+    Callable[[LearnerState, Transition, HiddenStates, HiddenStates], Tuple[LearnerState, Metrics]],
```

Comment:
This is annoying but not sure there is a way around it, seeing as you use it above can you store it in a variable please and use that instead